File: Chap6-GraphicsViewFramework/6-12FlappyBirdQt/scene.cpp
#include "scene.h"
#include <QGraphicsSceneMouseEvent>
#include <QKeyEvent>
#include <QDebug>

Scene::Scene(QObject *parent) : QGraphicsScene(parent),
    gameOn(false),score(0),
    bestScore(0)
{
    gameOverPix = nullptr;
    scoreTextItem = nullptr;
    setUpPillarTimer();
}

void Scene::addBird()
{
    bird = new BirdItem(QPixmap(":/images/bird_blue_up.png"));
    addItem(bird);
}

void Scene::startGame()
{
    //Bird
    bird->startFlying();
    //Pillars
    if(!pillarTimer->isActive()){
        cleanPillars();
        setGameOn(true);
        setScore(0);
        hideGameOverGraphics();
        pillarTimer->start(1000);
    }
}

void Scene::setUpPillarTimer()
{
    pillarTimer = new QTimer(this);
    connect(pillarTimer,&QTimer::timeout,[=](){


        PillarItem * pillarItem = new PillarItem();
        connect(pillarItem,&PillarItem::collideFail,[=](){
            pillarTimer->stop();
            freezeBirdAndPillarsInPlace();

            setGameOn(false);
            showGameOverGraphics();
        });

        addItem(pillarItem);

    });

    //pillarTimer->start(1000);

}

void Scene::freezeBirdAndPillarsInPlace()
{
    //Freeze bird
    bird->freezeInPlace();

    //Freeze pillars
    QList<QGraphicsItem *> sceneItems = items();
    foreach (QGraphicsItem *item, sceneItems) {
        PillarItem * pillar = dynamic_cast<PillarItem *>(item);
        if( pillar){
            pillar->freezeInPlace();
        }
    }
}

void Scene::setScore(int value)
{
    score = value;
}

bool Scene::getGameOn() const
{
    return gameOn;
}

void Scene::setGameOn(bool value)
{
    gameOn = value;
}

void Scene::incrementScore()
{
    score++;
    if(score > bestScore)
        bestScore = score;
    qDebug() << "Score : " << score << " Best Score : " << bestScore;

}

void Scene::keyPressEvent(QKeyEvent *event)
{
    if(event->key() == Qt::Key_Space){
        if(gameOn){
            bird->shootUp();
        }
    }
    QGraphicsScene::keyPressEvent(event);
}

void Scene::mousePressEvent(QGraphicsSceneMouseEvent *event)
{
    if(event->button() == Qt::LeftButton){
        if(gameOn){
            bird->shootUp();
        }
    }
    QGraphicsScene::mousePressEvent(event);
}

void Scene::showGameOverGraphics()
{
    gameOverPix = new QGraphicsPixmapItem(QPixmap(":/images/game_over_red.png"));
    addItem(gameOverPix);
    gameOverPix->setPos(QPointF(0,0) - QPointF(gameOverPix->boundingRect().width()/2,
                                               gameOverPix->boundingRect().height()/2));

    scoreTextItem = new QGraphicsTextItem();

    QString htmlString = "<p> Score : " + QString::number(score) + " </p>"
            + "<p> Best Score : " + QString::number(bestScore) + "</p>";

    QFont mFont("Consolas", 30, QFont::Bold);

    scoreTextItem->setHtml(htmlString);
    scoreTextItem->setFont(mFont);
    scoreTextItem->setDefaultTextColor(Qt::yellow);
    addItem(scoreTextItem);

    scoreTextItem->setPos(QPointF(0,0) - QPointF(scoreTextItem->boundingRect().width()/2,
                                                 -gameOverPix->boundingRect().height()/2));

}

void Scene::hideGameOverGraphics()
{
    if(gameOverPix){
        removeItem(gameOverPix);
        delete gameOverPix;
        gameOverPix= nullptr;
    }
    if(scoreTextItem){
        removeItem(scoreTextItem);
        delete  scoreTextItem;
        scoreTextItem = nullptr;
    }
}

void Scene::cleanPillars()
{
    QList<QGraphicsItem *> sceneItems = items();
    foreach (QGraphicsItem *item, sceneItems) {
        PillarItem * pillar = dynamic_cast<PillarItem *>(item);
        if( pillar){
            removeItem(pillar);
            delete pillar;
        }
    }
}

File: Chap6-GraphicsViewFramework/6-12FlappyBirdQt/scene.py
from PySide6.QtCore import QObject, QTimer, Qt
from PySide6.QtGui import QPixmap, QFont
from PySide6.QtWidgets import (QGraphicsScene, QGraphicsPixmapItem, 
                              QGraphicsTextItem)

class Scene(QGraphicsScene):
    def __init__(self, parent=None):
        super().__init__(parent)
        
        # Initialize variables
        self.game_on = False
        self.score = 0
        self.best_score = 0
        self.game_over_pix = None
        self.score_text_item = None
        self.bird = None
        
        # Set up pillar timer
        self.setup_pillar_timer()
    
    def add_bird(self):
        from birditem import BirdItem
        self.bird = BirdItem(QPixmap(":/images/bird_blue_up.png"))
        self.addItem(self.bird)
    
    def start_game(self):
        # Start bird flying
        self.bird.start_flying()
        
        # Handle pillars
        if not self.pillar_timer.isActive():
            self.clean_pillars()
            self.set_game_on(True)
            self.set_score(0)
            self.hide_game_over_graphics()
            # Add a 1.5 second delay before first pillar and more time between pillars
            self.game_start_delay = QTimer(self)
            self.game_start_delay.setSingleShot(True)
            self.game_start_delay.timeout.connect(lambda: self.pillar_timer.start(1800))
            self.game_start_delay.start(1500)
    
    def setup_pillar_timer(self):
        self.pillar_timer = QTimer(self)
        
        def create_pillar():
            from pillaritem import PillarItem
            pillar_item = PillarItem()
            pillar_item.collide_fail.connect(self.on_collision)
            self.addItem(pillar_item)
        
        self.pillar_timer.timeout.connect(create_pillar)
    
    def on_collision(self):
        self.pillar_timer.stop()
        self.freeze_bird_and_pillars_in_place()
        self.set_game_on(False)
        self.show_game_over_graphics()
    
    def freeze_bird_and_pillars_in_place(self):
        # Freeze bird
        self.bird.freeze_in_place()
        
        # Freeze pillars
        from pillaritem import PillarItem
        for item in self.items():
            if isinstance(item, PillarItem):
                item.freeze_in_place()
    
    def set_score(self, value):
        self.score = value
    
    def get_game_on(self):
        return self.game_on
    
    def set_game_on(self, value):
        self.game_on = value
    
    def increment_score(self):
        self.score += 1
        if self.score > self.best_score:
            self.best_score = self.score
        print(f"Score: {self.score} Best Score: {self.best_score}")
    
    def keyPressEvent(self, event):
        from PySide6.QtCore import Qt
        if event.key() == Qt.Key.Key_Space:
            if self.game_on:
                self.bird.shoot_up()
        super().keyPressEvent(event)
    
    def mousePressEvent(self, event):
        from PySide6.QtCore import Qt
        if event.button() == Qt.MouseButton.LeftButton:
            if self.game_on:
                self.bird.shoot_up()
        super().mousePressEvent(event)
    
    def show_game_over_graphics(self):
        # Add game over pixmap
        self.game_over_pix = QGraphicsPixmapItem(QPixmap(":/images/game_over_red.png"))
        self.addItem(self.game_over_pix)
        self.game_over_pix.setPos(
            0 - self.game_over_pix.boundingRect().width()/2, 
            0 - self.game_over_pix.boundingRect().height()/2
        )
        
        # Add score text
        self.score_text_item = QGraphicsTextItem()
        
        html_string = f"<p>Score: {self.score}</p><p>Best Score: {self.best_score}</p>"
        font = QFont("Consolas", 30, QFont.Bold)
        
        self.score_text_item.setHtml(html_string)
        self.score_text_item.setFont(font)
        from PySide6.QtCore import Qt
        self.score_text_item.setDefaultTextColor(Qt.yellow)
        self.addItem(self.score_text_item)
        
        self.score_text_item.setPos(
            0 - self.score_text_item.boundingRect().width()/2, 
            0 - self.game_over_pix.boundingRect().height()/2
        )
    
    def hide_game_over_graphics(self):
        if self.game_over_pix:
            self.removeItem(self.game_over_pix)
            self.game_over_pix = None
        
        if self.score_text_item:
            self.removeItem(self.score_text_item)
            self.score_text_item = None
    
    def clean_pillars(self):
        from pillaritem import PillarItem
        for item in list(self.items()):  # Make a copy to avoid modification during iteration
            if isinstance(item, PillarItem):
                self.removeItem(item)